### PR TITLE
Find number of columns and column placements

### PR DESCRIPTION
**Pull request by anbergem**

### CODE REVIEW

File: image_to_table/number_of_columns.py [hunk 1]
```diff
@@ -6,56 +6,70 @@
 import opencv_wrapper as cvw
 import scipy.ndimage as snd
 
+plt.set_cmap("gray")
```
Comment (anbergem) [inline image_to_table/number_of_columns.py]:
```suggestion
```

File: image_to_table/number_of_columns.py [hunk 2]
```diff
@@ -6,56 +6,70 @@
...
+    return image, first_white, last_white
+
+
+def find_column_placement(sum_image, num_columns):
+    thresh_sum = cvw.threshold_binary(sum_image, 1)
+    thresh_sum = thresh_sum[0]  # 1-D
+    clipped, first_white, _ = clip(thresh_sum)
+
+    is_zero = np.where(clipped == 0)[0]
+    column_breaks = sorted(consecutive(is_zero), key=len)
```
Comment (vegarsti) [inline image_to_table/number_of_columns.py:57]: 👌

File: image_to_table/number_of_columns.py [hunk 3]
```diff
@@ -6,56 +6,70 @@
...
+
+    while len(column_breaks) + 1 > num_columns:
+        clipped[column_breaks.pop(0)] = 255
+
+    return list(map(lambda x: first_white + x[-1], column_breaks))
+
+
+# https://stackoverflow.com/questions/7352684/how-to-find-the-groups-of-consecutive-elements-from-an-array-in-numpy
+def consecutive(data, stepsize=1):
+    return np.split(data, np.where(np.diff(data) != stepsize)[0] + 1)
```
Comment (vegarsti) [inline image_to_table/number_of_columns.py:71]: cool!

File: image_to_table/number_of_columns.py [hunk 4]
```diff
@@ -6,56 +6,70 @@
...
-    dilated = dilated[first_black:last_black]
-    diffed = np.diff(dilated)
+    gray = cvw.bgr2gray(image)
+
+    otsu = cvw.threshold_otsu(gray, inverse=True)
+
+    x_axis_sum = np.sum(otsu, axis=0).astype(np.float64)
+
+    # Expand to 2-D for image operations
+    sum_image = np.expand_dims(x_axis_sum, axis=0)
```
Comment (vegarsti) [inline image_to_table/number_of_columns.py]: maybe change `sum_image` to `expanded` or something?

File: image_to_table/number_of_columns.py [hunk 5]
```diff
@@ -6,56 +6,70 @@
...
     num_changes = np.count_nonzero(diffed)
     num_columns = (num_changes + 2) // 2
+
     return num_columns
+
+
+def clip(image):
+    first_white = np.argmax(image)
+    last_white = image.shape[0] - np.argmax(image[::-1])
+    image = image[first_white:last_white]
```
Comment (vegarsti) [inline image_to_table/number_of_columns.py:47]: Will this change the original image in-place as well? Don't think that's a problem, but curious

File: image_to_table/number_of_columns.py [hunk 6]
```diff
@@ -6,56 +6,70 @@
...
+def find_column_placement(sum_image, num_columns):
+    thresh_sum = cvw.threshold_binary(sum_image, 1)
+    thresh_sum = thresh_sum[0]  # 1-D
+    clipped, first_white, _ = clip(thresh_sum)
+
+    is_zero = np.where(clipped == 0)[0]
+    column_breaks = sorted(consecutive(is_zero), key=len)
+
+    while len(column_breaks) + 1 > num_columns:
+        clipped[column_breaks.pop(0)] = 255
```
Comment (vegarsti) [inline image_to_table/number_of_columns.py]: Should we add a comment here about what or why we do this? Or possibly extract into a function?

**Review by vegarsti — APPROVED** [review]
This is really great! This should make us really close to connecting it to the other parts!

A minor aesthetic thing: Some of the vertical spacing felt a bit haphazardly placed. I think it's hard to know when to add an empty line and when not to, myself. I felt like there were a bit too many empty lines. But not sure exactly which lines I would have removed.. E.g. I think you always have a space before the return statement. I think I would have removed them. But again, never sure what to do here.

Also, we can check soon how much speedup it may give if we run the number and placement functions concurrently. Can do that by

```python
with concurrency.futures.ThreadPoolWorker(max_workers=2) as e:
    num_columns_future = e.submit(fn=find_number_of_columns, sum_image=sum_image)
    column_placement_future = e.submit(fn=find_column_placement, sum_image=sum_image, num_columns=num_columns)
# with block is done after both tasks are done
num_columns = num_columns_future.result()
column_placement = column_placement_future.result()
```

File: image_to_table/number_of_columns.py (hunk 5 above)
Comment (anbergem) [reply image_to_table/number_of_columns.py:47]: Yes. Avoid making a copy 😄

**Review by anbergem — COMMENTED** [review]
Remove old code

**Comment by anbergem** [comment]
I agree with the vertical spacing thing... I am always very unsure. I generally to like a line break before return, and otherwise try to group things logically, separated by a line break. But... I don't always achieve anything with it, or manage to group things logically... But 15 lines of code with noe line break is a bit harsh on the eyes, IMO.

My thought was that these two methods run in a single thread next to a thread running google vision. `find_column_placement` needs the number of columns when merging, so it has to come after `find_number_of_columns`.

**Comment by vegarsti** [comment]
> 15 lines of code with noe line break is a bit harsh on the eyes, IMO.
true!

> My thought was that these two methods run in a single thread next to a thread running google vision. find_column_placement needs the number of columns when merging, so it has to come after find_number_of_columns.

Ah of course. But good, running it alongside the vision code definitely
helps a lot!

**Pull request merged**